qa/L0_python_api/test_kserve.py

```diff
@@ -108,9 +115,63 @@ def test_wrong_http_parameters(self):
             Metrics.Options(thread_count="ten")
 
 
+class TestRestrictedFeatureOptions:
+    def test_correct_rf_parameters(self):
+        # Directly test feature groups
+        correct_feature_group = FeatureGroup(
+            key="key", value="val", features=[Feature.HEALTH, Feature.METADATA]
+        )
+
+        rf = RestrictedFeatures(groups=[correct_feature_group])
+        rf.create_feature_group(
+            key="new-key", value="new-val", features=[Feature.INFERENCE]
+        )
+
+    def test_wrong_rf_parameters(self):
+        with pytest.raises(AttributeError):
+            Feature.health  # Needs to be HEALTH
+        with pytest.raises(AttributeError):
+            Feature.infer  # Needs to be INFERENCE
+
+        rf = RestrictedFeatures()
+        # Features List needs to be an element from tritonfrontend.Feature
+        with pytest.raises(tritonserver.InvalidArgumentError):
+            rf.create_feature_group(key="", value="", features=["health"])
+
+        # key and value need to be of type string
+        with pytest.raises(Exception):
+            rf.create_feature_group(
+                key=42, value="Secret to the Universe", features=[Feature.HEALTH]
+            )
+        with pytest.raises(Exception):
+            rf.create_feature_group(key="", value=123, features=[Feature.HEALTH])
+
+        # Test collision of Features among individual Feature Groups
+        with pytest.raises(tritonserver.InvalidArgumentError):
+            feature_group = FeatureGroup(
+                key="key", value="val", features=[Feature.METADATA, Feature.HEALTH]
+            )
+
+            rf = RestrictedFeatures(groups=[feature_group])
+            rf.create_feature_group(key="key2", value="val", features=[Feature.HEALTH])
+
+        with pytest.raises(tritonserver.InvalidArgumentError):
+            rf = RestrictedFeatures(groups=[feature_group])
+            rf.create_feature_group(
+                key="key", value="val", features=[Feature.METADATA, Feature.HEALTH]
+            )
+            rf.create_feature_group(key="key2", value="val", features=[Feature.HEALTH])
+
+
 HTTP_ARGS = (KServeHttp, httpclient, "localhost:8000")  # Default HTTP args
 GRPC_ARGS = (KServeGrpc, grpcclient, "localhost:8001")  # Default GRPC args
 METRICS_ARGS = (Metrics, "localhost:8002")  # Default Metrics args
+RESTRICTED_FEATURE_ARG = RestrictedFeatures()
```

Comment:
Is this the pytest way to set test suit level parameters? Can you point me to the guide on that?

Reply:
The only suit level parameter initialization and testing I could find was through fixtures and since we haven't used fixtures throughout this test, for the sake of consistency I moved the `RestrictedFeatures` object to inside the only function that was testing it. If we would like to fixtures, I can look into that as well.